### PR TITLE
Fix/sidecar migration

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR fixes two issues failed migration for models with sidecar applications;
- https://bugs.launchpad.net/juju/+bug/2009566
- https://bugs.launchpad.net/juju/+bug/2008744

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
juju deploy prometheus-k8s --trust

juju deploy cs:~juju/mariadb-k8s-3

juju migrate k1:t1 k2
Migration started with ID "d2254384-b25c-492b-8e69-798f66e2821c:0"

```

## Documentation changes

No

## Bug reference

- https://bugs.launchpad.net/juju/+bug/2009566
- https://bugs.launchpad.net/juju/+bug/2008744
-

### CODE REVIEW

File: caas/kubernetes/provider/resources/statefulset.go [hunk 1]
```diff
@@ -48,28 +49,56 @@ func (ss *StatefulSet) ID() ID {
...
+	result, err = api.Create(ctx, &ss.StatefulSet, metav1.CreateOptions{
+		FieldManager: JujuFieldManager,
+	})
+	if k8serrors.IsAlreadyExists(err) {
+		// Continue to patch.
+	} else if err != nil {
+		return errors.Trace(err)
+	} else {
+		return nil
+	}
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/statefulset.go:70]: We'd normally write this as
```
if err == nil || !k8serrors.IsAlreadyExists(err) {
  return errors.Trace(err)
}
...
do patch
```

File: caas/kubernetes/provider/resources/statefulset.go [hunk 2]
```diff
@@ -48,28 +49,56 @@ func (ss *StatefulSet) ID() ID {
...
+
+	existing.SetAnnotations(ss.GetAnnotations())
+	existing.Spec.Replicas = ss.Spec.Replicas
+	existing.Spec.UpdateStrategy = ss.Spec.UpdateStrategy
+	existing.Spec.Template.Spec.Volumes = ss.Spec.Template.Spec.Volumes
+	existing.Spec.Template.SetAnnotations(ss.Spec.Template.GetAnnotations())
+	existing.Spec.Template.Spec.Containers = ss.Spec.Template.Spec.Containers
+	existing.Spec.Template.Spec.ServiceAccountName = ss.Spec.Template.Spec.ServiceAccountName
+	existing.Spec.Template.Spec.AutomountServiceAccountToken = ss.Spec.Template.Spec.AutomountServiceAccountToken
+	// NB: we can't update the Spec.ServiceName as it is immutable.
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/statefulset.go:85]: This change makes statefulset updates different to all the other k8s resources juju manages. Everything currently uses the same pattern:
1. patch
2. if not found, create

Can we simply alter the existing logic (for everything) to check the error from patching and if it's "ErrConflict", then refresh the resource and try again?

**Review by wallyworld — COMMENTED** [review]
A question about the fix for statefulsets - I don't like how now statefulsets are treated differently to everything else. Can we adopt an approach that keeps the update logic consistent for all k8s resources juju manages.

**Comment by ycliuhw** [comment]
> A question about the fix for statefulsets - I don't like how now statefulsets are treated differently to everything else. Can we adopt an approach that keeps the update logic consistent for all k8s resources juju manages.

Statefulset is different from everything else. We do the similar thing for operator-style applications as well.

**Review by wallyworld — APPROVED** [review]

**Pull request merged**